src/main/scala/is/hail/expr/Relational.scala

```diff
@@ -345,6 +345,86 @@ object MatrixIR {
       keepF(mv, keep)
     })
   }
+
+  def collectColsByKey(typ: MatrixType): (MatrixType, MatrixValue => MatrixValue) = {
+    val newColValueType = TStruct(typ.colValueStruct.fields.map(f => f.copy(typ = TArray(f.typ, required = true))))
+    val newColType = typ.colKeyStruct ++ newColValueType
+    val newEntryType = TStruct(typ.entryType.fields.map(f => f.copy(typ = TArray(f.typ, required = true))))
+    val newMatrixType = typ.copyParts(colType = newColType, entryType = newEntryType)
+    val oldRVRowType = typ.rvRowType
+    val newRVRowType = newMatrixType.rvRowType
+    val oldEntryArrayType = typ.entryArrayType
+    val oldEntryType = typ.entryType
+    val localRowSize = newRVRowType.size
+
+    (newMatrixType, { mv =>
+      val colValMap: Map[Row, Array[Int]] = mv.colValues.value.map(_.asInstanceOf[Row]).zipWithIndex
+        .groupBy[Row]{ case (r, i) => typ.extractColKey(r)}
+        .mapValues{ _.map{ case (r, i) => i }.toArray}
```

Comment:
Given that `(r, i)` is right above, I'd simplify `_.map{ case (r, i) => i` to `_._2`. But I could see argument that writing it out is clearer, up to you. Move bracket to next line.

Reply:
I do prefer writing out the case statement, and having the two functionals be consistent.

I'm not sure how you want the line breaks.

Reply:
Cool. Suggesting line break before .map (and space before bracket on third line)
```
val colValMap: Map[Row, Array[Int]] = mv.colValues.value
  .map(_.asInstanceOf[Row]).zipWithIndex
  .groupBy[Row]{ case (r, i) => typ.extractColKey(r) }
  .mapValues{ _.map{ case (r, i) => i }.toArray }
```